tests/unit/cbprintf/main.c

```diff
@@ -0,0 +1,1056 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <ztest.h>
+#include <float.h>
+#include <limits.h>
+#include <math.h>
+#include <stdarg.h>
+#include <wctype.h>
+#include <stddef.h>
+#include <string.h>
+#include <sys/cbprintf.h>
+#include <sys/util.h>
+
+/* Unit testing doesn't use Kconfig, so if we're not building from
+ * sanitycheck force selection of all features.  If we are use flags
+ * to determine which features are desired.  Yes, this is a mess.
+ */
+#ifndef VIA_SANITYCHECK
+/* Set this to truthy to use libc's snprintf as external validation.
+ * This should be used with all options enabled.
+ */
+#define USE_LIBC 0
+#define CONFIG_CBPRINTF_COMPLETE 1
+#define CONFIG_CBPRINTF_FULL_INTEGRAL 1
+#define CONFIG_CBPRINTF_FP_SUPPORT 1
+#define CONFIG_CBPRINTF_FP_A_SUPPORT 1
+#define CONFIG_CBPRINTF_N_SPECIFIER 1
+#define CONFIG_CBPRINTF_LIBC_SUBSTS 1
+
+/* compensate for selects */
+#if (CONFIG_CBPRINTF_FP_A_SUPPORT - 0)
+#undef CONFIG_CBPRINTF_REDUCED_INTEGRAL
+#undef CONFIG_CBPRINTF_FULL_INTEGRAL
+#undef CONFIG_CBPRINTF_FP_SUPPORT
+#define CONFIG_CBPRINTF_FULL_INTEGRAL 1
+#define CONFIG_CBPRINTF_FP_SUPPORT 1
+#endif
+
+#if !(CONFIG_CBPRINTF_FP_SUPPORT - 0)
+#undef CONFIG_CBPRINTF_FP_SUPPORT
+#endif
+
+/* Convert choice selections to one defined flag */
+#if !(CONFIG_CBPRINTF_COMPLETE - 0)
+#ifdef CONFIG_CBPRINTF_FP_SUPPORT
+#error Inconsistent configuration
+#endif
+#undef CONFIG_CBPRINTF_COMPLETE
+#define CONFIG_CBPRINTF_NANO 1
+#endif
+#if !(CONFIG_CBPRINTF_FULL_INTEGRAL - 0)
+#undef CONFIG_CBPRINTF_FULL_INTEGRAL
+#define CONFIG_CBPRINTF_REDUCED_INTEGRAL 1
+#endif
+
+#else /* VIA_SANITYCHECK */
+#if (VIA_SANITYCHECK & 0x01) != 0
+#define CONFIG_CBPRINTF_FULL_INTEGRAL 1
+#else
+#define CONFIG_CBPRINTF_REDUCED_INTEGRAL 1
+#endif
+#if (VIA_SANITYCHECK & 0x02) != 0
+#define CONFIG_CBPRINTF_FP_SUPPORT 1
+#endif
+#if (VIA_SANITYCHECK & 0x04) != 0
+#define CONFIG_CBPRINTF_FP_A_SUPPORT 1
+#endif
+#if (VIA_SANITYCHECK & 0x08) != 0
+#define CONFIG_CBPRINTF_N_SPECIFIER 1
+#endif
+#if (VIA_SANITYCHECK & 0x40) != 0
+#define CONFIG_CBPRINTF_FP_ALWAYS_A 1
+#endif
+#if (VIA_SANITYCHECK & 0x80) != 0
+#define CONFIG_CBPRINTF_NANO 1
+#else /* 0x80 */
+#define CONFIG_CBPRINTF_COMPLETE 1
+#endif /* 0x80 */
+#if (VIA_SANITYCHECK & 0x100) != 0
+#define CONFIG_CBPRINTF_LIBC_SUBSTS 1
+#endif
+
+#endif /* VIA_SANITYCHECK */
+
+#include "../../../lib/os/cbprintf.c"
+
+#if defined(CONFIG_CBPRINTF_COMPLETE)
+#include "../../../lib/os/cbprintf_complete.c"
+#elif defined(CONFIG_CBPRINTF_NANO)
+#include "../../../lib/os/cbprintf_nano.c"
+#endif
+
+/* We can't determine at build-time whether int is 64-bit, so assume
```

Comment:
> We can't determine at build-time whether int is 64-bit, so assume

how about `BUILD_ASSERT(sizeof(int) < 8)`?

The below casting where you do stuff like
```
	static const unsigned int pfx_val = (unsigned int)0x7b6b5b4b3b2b1b0b;
```

Are you sure that this is defined behavior?
This is generating a CI lintian warning BTW

Reply:
I'm not trying to test whether sizeof(int) < 8, I'm trying to support sizeof(int) == 8, in the belief that someday somebody might want Zephyr to run on some DSP-like ILP64 system.

The diagnostic is

    WARNING:TYPECAST_INT_CONSTANT: Unnecessary typecast of c90 int constant

but in fact it is necessary to cast: if the cast is removed the assignment produces warnings that the 64-bit constant was truncated.

There's nothing wrong with the code (there's nothing wrong with the cast), it's just checkpatch making assumptions that aren't relevant in this special case.

Similarly the test is verifying that `%Ld` is properly handled by the converter: i.e. it is detected and passed through unchanged.  This again is checkpatch making assumptions that any string containing an invalid format specifier in any context must have been meant to be a format specifier:

```
-:4889: WARNING:PRINTF_L: %Ld is non-standard C, use %lld
#4889: FILE: tests/unit/cbprintf/main.c:445:
+	zassert_equal(strncmp("/%Ld/", buf, rc), 0, NULL);
```

Here it's complaining that it's being passed to `strncmp`.  checkpatch is wrong to complain about this.

If you prefer we can eliminate these complaints by (a) not being prepared to support ILP64, and/or (b) not testing handling of undefined format combinations.  But I'd rather just ignore checkpatch's whingeing.